File: 2229-maximum-fruits-harvested-after-at-most-k-steps/maximum-fruits-harvested-after-at-most-k-steps.py
class Solution:
    def maxTotalFruits(self, fruits: list[list[int]], startPos: int, k: int) -> int:
        res = 0
        left = 0
        curr_sum = 0

        for right in range(len(fruits)):
            curr_sum += fruits[right][1]

            # shrink window from the left if invalid
            while left <= right and not self.isValidRange(fruits[left][0], fruits[right][0], startPos, k):
                curr_sum -= fruits[left][1]
                left += 1

            res = max(res, curr_sum)

        return res

    def isValidRange(self, leftPos: int, rightPos: int, startPos: int, k: int) -> bool:
        if rightPos <= startPos:
            return startPos - leftPos <= k
        elif leftPos >= startPos:
            return rightPos - startPos <= k
        else:
            left = startPos - leftPos
            right = rightPos - startPos
            return (left * 2 + right <= k) if left <= right else (right * 2 + left <= k)

File: 2229-maximum-fruits-harvested-after-at-most-k-steps/maximum-fruits-harvested-after-at-most-k-steps.go
func maxTotalFruits(fruits [][]int, startPos int, k int) int {
	res := 0
	sum := 0
	left := 0

	for right := 0; right < len(fruits); right++ {
		sum += fruits[right][1]

		// Shrink window from the left while it's invalid
		for left <= right && !isValidRange(fruits[left][0], fruits[right][0], startPos, k) {
			sum -= fruits[left][1]
			left++
		}

		if sum > res {
			res = sum
		}
	}

	return res
}

func isValidRange(leftPos, rightPos, startPos, k int) bool {
	if rightPos <= startPos {
		return startPos-leftPos <= k
	} else if leftPos >= startPos {
		return rightPos-startPos <= k
	} else {
		left := startPos - leftPos
		right := rightPos - startPos
		if left <= right {
			return left*2 + right <= k
		}
		return right*2 + left <= k
	}
}
